src/GafferScene/IECoreScenePreview/PrimitiveAlgo.cpp

```diff
@@ -0,0 +1,1188 @@
+				)
+				{
+					return DataTraits< std::vector<ValueType> >::DataType::staticTypeId();
+				}
+			}
+			return IECore::InvalidTypeId;
+		}
+	);
+}
+
+bool interpolationMatches(
+	IECoreScene::TypeId primType, PrimitiveVariable::Interpolation a, PrimitiveVariable::Interpolation b
+)
+{
+	if( a == b )
+	{
+		return true;
+	}
+
+	if( primType == IECoreScene::MeshPrimitiveTypeId )
+	{
+		auto isVertex = []( PrimitiveVariable::Interpolation x) {
+			return x == PrimitiveVariable::Vertex || x == PrimitiveVariable::Varying;
+		};
+		return isVertex( a ) && isVertex( b );
+	}
+	else if( primType == IECoreScene::MeshPrimitiveTypeId )
+	{
+		auto isVarying = []( PrimitiveVariable::Interpolation x) {
+			return x == PrimitiveVariable::Varying || x == PrimitiveVariable::FaceVarying;
+		};
+		return isVarying( a ) && isVarying( b );
+	}
+	else
+	{
+		auto isVertex = []( PrimitiveVariable::Interpolation x) {
+			return x == PrimitiveVariable::Vertex || x == PrimitiveVariable::Varying || x == PrimitiveVariable::FaceVarying;
+		};
+		return isVertex( a ) && isVertex( b );
+	}
+}
+
+// Set up indices on the destination matching the source indices ( includes handling converting interpolations ).
+// Note that this only handles interpolations used by mergePrimitives ( ie. only promotion to more specific
+// interpolations, like Vertex -> FaceVarying, but it cann't do averaging ).
+void copyIndices(
+	const std::vector<int> *sourceIndices, int *destIndices,
+	IECoreScene::TypeId primTypeId,
+	PrimitiveVariable::Interpolation sourceInterp, PrimitiveVariable::Interpolation destInterp,
+	size_t numIndices, size_t dataStart,
+	const Primitive *sourcePrim
+)
+{
+	// Helper function that translates from an index in the source data to an index in the destination
+	// data, based on the data offset, and the source indices ( if present )
+	auto translateIndex = [sourceIndices, dataStart]( int j ){
+		return sourceIndices ? dataStart + (*sourceIndices)[j] : dataStart + j;
+	};
+
+	if( interpolationMatches( primTypeId, sourceInterp, destInterp ) )
+	{
+		// If the interpolation hasn't changed, we don't need to anything special, just translate
+		// each index.
+		for( size_t j = 0; j < numIndices; j++ )
+		{
+			*(destIndices++) = translateIndex( j );
+		}
+	}
+	else if( sourceInterp == PrimitiveVariable::Constant )
+	{
+		// Constant variables aren't stored as vectors, so they can't have been indexed to start,
+		// just set all the output indices to the one element of output data.
+		for( size_t j = 0; j < numIndices; j++ )
+		{
+			*(destIndices++) = dataStart;
+		}
+	}
+	else if( sourceInterp == PrimitiveVariable::Uniform )
+	{
+		if( primTypeId == IECoreScene::MeshPrimitiveTypeId )
+		{
+			// On a mesh, if you combine a Uniform with anything it doesn't match, then it gets
+			// promoted to FaceVarying.
+			assert( destInterp == PrimitiveVariable::FaceVarying );
+
+			const MeshPrimitive *sourceMesh = IECore::runTimeCast< const MeshPrimitive >( sourcePrim );
```

Comment:
So, this change could definitely be made, and it's not a problem for correctness or performance, but it doesn't feel right to me ... the primTypeId isn't just related to the source primitive, it determines the meanings of all the buffers that are being passed here. For example, if the prim type is Mesh, and sourceInterp is Varying, and the destInterp is Vertex, then you can copy straight across without resampling, and you don't look at sourcePrim.

It's true that we shouldn't be using unchecked runTimeCasts, but I would prefer to just replace these with static_casts. This function already requires that the input all be of matching primType ... if the calling code has gotten confused about that, the allocations for sourceIndices and destIndices won't match and we'll just crash anyway.

However, since sourcePrim is needed for some of the cases where we do need to resample, we can get the prim type from there and save an argument. I don't have a problem doing this, just thought I'd double check if it feels weird to you.